collectors/node/metrics.go

```diff
@@ -15,225 +15,105 @@
-	TxDropped uint64 `json:"network.{{.Name}}.out.dropped"`
-	RxErrors  uint64 `json:"network.{{.Name}}.in.errors"`
-	TxErrors  uint64 `json:"network.{{.Name}}.out.errors"`
-}
-
-func getNodeMetrics() (nodeMetrics, error) {
-	l := getLoadAvg()
-	cpuStatePcts := calculatePcts(getCPUTimes())
-	m := getMemory()
-	s := getSwap()
-
-	return nodeMetrics{
-		Uptime:       getUptime(),
-		ProcessCount: getProcessCount(),
-
-		NumCores:  getNumCores(),
-		Load1Min:  l.Load1,
-		Load5Min:  l.Load5,
-		Load15Min: l.Load15,
-
-		// Percentages are calculated between the last poll of the CPU times
-		// and the current times, as stored in lastCPU. 100.00 => 100.00%
-		CPUTotalPct:  cpuStatePcts.User + cpuStatePcts.System,
-		CPUUserPct:   cpuStatePcts.User,
-		CPUSystemPct: cpuStatePcts.System,
-		CPUIdlePct:   cpuStatePcts.Idle,
-		CPUWaitPct:   cpuStatePcts.Iowait,
-
-		MemTotalBytes:   m.Total,
-		MemFreeBytes:    m.Free,
-		MemBuffersBytes: m.Buffers,
-		MemCachedBytes:  m.Cached,
-
-		SwapTotalBytes: s.Total,
-		SwapFreeBytes:  s.Free,
-		SwapUsedBytes:  s.Used,
-
-		Filesystems:       getFilesystems(),
-		NetworkInterfaces: getNetworkInterfaces(),
-	}, nil
-}
+	"github.com/dcos/dcos-metrics/producers"
+)
 
-// -- helpers
+const (
+	// Unit constants
+	COUNT = "count"
+	BYTES = "bytes"
+
+	/* Metric name constants */
+	UPTIME = "uptime"
+
+	/* process.<namespace> */
+	PROCESS_COUNT = "process.count"
+
+	/* load.<namespace> */
+	LOAD_1MIN  = "load.1min"
+	LOAD_5MIN  = "load.5min"
+	LOAD_15MIN = "load.15min"
+
+	/* load.<namespace> */
+	CPU_CORES  = "cpu.cores"
+	CPU_TOTAL  = "cpu.total"
+	CPU_USER   = "cpu.user"
+	CPU_SYSTEM = "cpu.system"
+	CPU_IDLE   = "cpu.idle"
+	CPU_WAIT   = "cpu.wait"
+
+	/* filesystem.<namespace> */
+	FS_CAP_TOTAL   = "filesystem.capacity.total"
+	FS_CAP_USED    = "filesystem.capacity.used"
+	FS_CAP_FREE    = "filesystem.capacity.free"
+	FS_INODE_TOTAL = "filesystem.inode.total"
+	FS_INODE_USED  = "filesystem.inode.used"
+	FS_INODE_FREE  = "filesystem.inode.free"
+
+	/* network.<namespace> */
+	NET_IN          = "network.in"
+	NET_OUT         = "network.out"
+	NET_IN_PACKETS  = "network.in.packets"
+	NET_OUT_PACKETS = "network.out.packets"
+	NET_IN_DROPPED  = "network_in_dropped"
+	NET_OUT_DROPPED = "network_out_dropped"
+	NET_IN_ERRORS   = "network_in_errors"
+	NET_OUT_ERRORS  = "network_out_errors"
+
+	/* memory.<namespace> */
+	MEM_TOTAL   = "memory.total"
+	MEM_FREE    = "memory.free"
+	MEM_BUFFERS = "memory.buffers"
+	MEM_CACHED  = "memory.cached"
+
+	/* swap.<namespace> */
+	SWAP_TOTAL = "swap.total"
+	SWAP_FREE  = "swap.free"
+	SWAP_USED  = "swap.used"
+)
 
-type lastCPUTimes struct {
-	sync.Mutex
-	times cpu.TimesStat
+type nodeCollector struct {
+	datapoints []producers.Datapoint
 }
 
-var lastCPU lastCPUTimes
-
-func init() {
-	t, _ := cpu.Times(false) // get totals, not per-cpu stats
-	lastCPU.Lock()
-	lastCPU.times = t[0]
-	lastCPU.Unlock()
+type nodeMetricPoller interface {
```

Comment:
I am using it in an effort to set a precedence on how to add new metrics to this codebase later. If myself or another developer adds code to this, I want there to be a strictly defined pattern on how that is done. Though an interface may seem overly complex for this, since we can easily just add getMeomry() or getThisMetric() functions in lieu of them, the interface does provide a roadmap on how to add code to this project. 

If we use utility functions, we'll end up with different naming mechanisms, polling methods,  etc. Using an interface means every metric gathered is done so in the same way across the codebase. This makes it efficient to onboard new devs or to remind old devs how to add to this codebase if they haven't in a while. 

The extra 4 lines of code it took us to define this interface really define a roadmap, making this code less complex, even if on the surface it might not seem that way.